### ISSUE TITLE
Segfault if remote WASM load fails

### ISSUE THREAD

**Issue by mandarjog** [envoyproxy/envoy-wasm#436]
Running with latest release-1.5 proxy sha: ee4327fea8bfbde96f4516fd91da265f2475e91e
```
[2020-02-26 21:03:52.970][10627079][debug][init] [external/envoy/source/common/init/target_impl.cc:15] init manager Server initializing target RemoteAsyncDataProvider
[2020-02-26 21:03:52.970][10627079][debug][config] [external/envoy/source/common/config/remote_data_fetcher.cc:35] fetch remote data from [uri = https://github.com/istio/proxy/blob/release-1.5/extensions/stats/plugin.wasm?raw=true]: start
[2020-02-26 21:03:52.970][10627079][debug][router] [external/envoy/source/common/router/router.cc:474] [C0][S16998448541629250160] cluster 'com_github' match for URL '/istio/proxy/blob/release-1.5/extensions/stats/plugin.wasm?raw=true'
[2020-02-26 21:03:52.970][10627079][debug][router] [external/envoy/source/common/router/router.cc:614] [C0][S16998448541629250160] router decoding headers:
':path', '/istio/proxy/blob/release-1.5/extensions/stats/plugin.wasm?raw=true'
':authority', 'github.com'
':method', 'GET'
':scheme', 'http'
'x-envoy-internal', 'true'
'x-forwarded-for', '192.168.86.194'
'x-envoy-expected-rq-timeout-ms', '5000'

[2020-02-26 21:03:52.971][10627079][debug][pool] [external/envoy/source/common/http/http1/conn_pool.cc:95] creating a new connection
[2020-02-26 21:03:52.971][10627079][debug][client] [external/envoy/source/common/http/codec_client.cc:34] [C0] connecting
[2020-02-26 21:03:52.971][10627079][debug][connection] [external/envoy/source/common/network/connection_impl.cc:691] [C0] connecting to 192.30.255.112:443
[2020-02-26 21:03:52.971][10627079][debug][connection] [external/envoy/source/common/network/connection_impl.cc:700] [C0] connection in progress
[2020-02-26 21:03:52.971][10627079][debug][pool] [external/envoy/source/common/http/conn_pool_base.cc:55] queueing request due to no available connections
[2020-02-26 21:03:53.001][10627079][debug][connection] [external/envoy/source/common/network/connection_impl.cc:563] [C0] connected
[2020-02-26 21:03:53.001][10627079][debug][client] [external/envoy/source/common/http/codec_client.cc:72] [C0] connected
[2020-02-26 21:03:53.001][10627079][debug][pool] [external/envoy/source/common/http/http1/conn_pool.cc:244] [C0] attaching to next request
[2020-02-26 21:03:53.001][10627079][debug][router] [external/envoy/source/common/router/router.cc:1711] [C0][S16998448541629250160] pool ready
[2020-02-26 21:03:53.040][10627079][debug][connection] [external/envoy/source/common/network/connection_impl.cc:531] [C0] remote close
[2020-02-26 21:03:53.040][10627079][debug][connection] [external/envoy/source/common/network/connection_impl.cc:192] [C0] closing socket: 0
[2020-02-26 21:03:53.041][10627079][debug][client] [external/envoy/source/common/http/codec_client.cc:91] [C0] disconnect. resetting 1 pending requests
[2020-02-26 21:03:53.041][10627079][debug][client] [external/envoy/source/common/http/codec_client.cc:114] [C0] request reset
[2020-02-26 21:03:53.041][10627079][debug][router] [external/envoy/source/common/router/router.cc:990] [C0][S16998448541629250160] upstream reset: reset reason connection termination
[2020-02-26 21:03:53.047][10627079][debug][http] [external/envoy/source/common/http/async_client_impl.cc:95] async http request response headers (end_stream=false):
':status', '503'
'content-length', '95'
'content-type', 'text/plain'

[2020-02-26 21:03:53.047][10627079][debug][config] [external/envoy/source/common/config/remote_data_fetcher.cc:62] fetch remote data [uri = https://github.com/istio/proxy/blob/release-1.5/extensions/stats/plugin.wasm?raw=true]: response status code 503
[2020-02-26 21:03:53.047][10627079][debug][config] [bazel-out/darwin-dbg/bin/external/envoy/source/common/config/_virtual_includes/datasource_lib/common/config/datasource.h:83] Failed to fetch remote data. Failure reason: 0
[2020-02-26 21:03:53.047][10627079][critical][main] [external/envoy/source/exe/terminate_handler.cc:13] std::terminate called! (possible uncaught exception, see trace)
[2020-02-26 21:03:53.047][10627079][critical][backtrace] [bazel-out/darwin-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[2020-02-26 21:03:53.047][10627079][critical][backtrace] [bazel-out/darwin-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 7e85a2a08d9a872e908ec229e35116931762ee00/1.13.1-dev/Modified/DEBUG/BoringSSL
[2020-02-26 21:03:53.047][10627079][critical][backtrace] [bazel-out/darwin-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Abort trap: 6, suspect faulting address 0x7fff7a3662c2
[2020-02-26 21:03:53.047][10627079][critical][backtrace] [bazel-out/darwin-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[2020-02-26 21:03:53.047][10627079][critical][backtrace] [bazel-out/darwin-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 7e85a2a08d9a872e908ec229e35116931762ee00/1.13.1-dev/Modified/DEBUG/BoringSSL
AsyncClient 0xa46bb10, stream_id_: 16998448541629250160
&stream_info_:
  StreamInfoImpl 0xa46bcb0, protocol_: 1, response_code_: 503, response_code_details_: upstream_reset_before_response_started{connection termination}, health_check_request_: 0, route_name_:
./run_test.sh: line 23: 53512 Abort trap: 6           ${BAZEL_BIN}/src/envoy/envoy -c ${WD}/testdata/server.yaml --concurrency 2 --allow-unknown-fields -l debug
```

**Comment by mandarjog** [comment]
@jplevyak @PiotrSikora

**Comment by PiotrSikora** [comment]
cc @yxue

**Comment by yxue** [comment]
The root cause is the REST api returns 503 and Envoy crashes. I had a PR to
add the retry policy and it was blocked by another PR.
https://github.com/envoyproxy/envoy/pull/9463

John and me had the discussion before about how to handle the 503 returned
by remote data source. The initial solution is throwing an exception but we
figured out that the exception won’t be caught correctly. The current
solution is just logging the failure and let Envoy crash. Hopefully adding
retry policy will make it more reliable.

**Comment by jplevyak** [comment]
Perhaps we should change it to throwing an exception.  At least that indicates that the issue is known. 

Looks like https://github.com/envoyproxy/envoy/pull/9346 is closed now.   Perhaps you can merge it into 9463 and try to upstream the combo ?

**Comment by yxue** [comment]
First I am going to merge the PRs related to data source retry policy.

For throwing exception, since we have already logged the error, do we still
need to throw the exception considering the behavior is unpredictable?
